diff --git a/components/driver/include/driver/i2s_types.h b/components/driver/include/driver/i2s_types.h
index 84cd39241b40eed7885378da73771d418d75c331..f09a6fc06831992c7819e04a3cdf4b98867b4f56 100644
--- a/components/driver/include/driver/i2s_types.h
+++ b/components/driver/include/driver/i2s_types.h
@@ -63,7 +63,11 @@ typedef struct {
                                   */
 } i2s_event_data_t;
 
+#ifdef __bindgen
+typedef void    *i2s_chan_handle_t; /*!< i2s channel handle, the control unit of the i2s driver*/
+#else
 typedef struct i2s_channel_t    *i2s_chan_handle_t; /*!< i2s channel handle, the control unit of the i2s driver*/
+#endif
 
 /**
  * @brief I2S event callback
