diff --git a/components/riscv/stdatomic.c b/components/riscv/stdatomic.c
index 9cae8ee518..015fb8fc01 100644
--- a/components/riscv/stdatomic.c
+++ b/components/riscv/stdatomic.c
@@ -37,6 +37,21 @@
     portEXIT_CRITICAL_NESTED(state); \
     } while (0)
 
+#define ATOMIC_LOAD(n, type) type __atomic_load_ ## n (const type* mem, int memorder) \
+{                                                   \
+    unsigned state = _ATOMIC_ENTER_CRITICAL();      \
+    type ret = *mem;                                \
+    _ATOMIC_EXIT_CRITICAL(state);                   \
+    return ret;                                     \
+}
+
+#define ATOMIC_STORE(n, type) void __atomic_store_ ## n (type* mem, type val, int memorder) \
+{                                                   \
+    unsigned state = _ATOMIC_ENTER_CRITICAL();      \
+    *mem = val;                                     \
+    _ATOMIC_EXIT_CRITICAL(state);                   \
+}
+
 #define ATOMIC_EXCHANGE(n, type) type __atomic_exchange_ ## n (type* mem, type val, int memorder) \
 {                                                   \
     unsigned state = _ATOMIC_ENTER_CRITICAL();      \
@@ -141,6 +156,16 @@
 
 #pragma GCC diagnostic ignored "-Wbuiltin-declaration-mismatch"
 
+ATOMIC_LOAD(1, uint8_t)
+ATOMIC_LOAD(2, uint16_t)
+ATOMIC_LOAD(4, uint32_t)
+ATOMIC_LOAD(8, uint64_t)
+
+ATOMIC_STORE(1, uint8_t)
+ATOMIC_STORE(2, uint16_t)
+ATOMIC_STORE(4, uint32_t)
+ATOMIC_STORE(8, uint64_t)
+
 ATOMIC_EXCHANGE(1, uint8_t)
 ATOMIC_EXCHANGE(2, uint16_t)
 ATOMIC_EXCHANGE(4, uint32_t)
